### PR TITLE
Features/ontouml2gufo relations

### PR DESCRIPTION
**Pull request by LucasBassetti**
This transformation addresses:

- [x] Relation stereotypes
- [x] Relation domain / rage
- [x] Relation cardinalities
- [x] Option to use name or id as uri

Example of output:

```
:benefits rdfs:subPropertyOf gufo:mediates.
:SocialRelationship rdfs:subClassOf gufo:Relator;
    rdf:type gufo:Kind.
:deliveredby rdf:type owl:ObjectProperty;
    rdfs:domain :LongTermHealthcareProvider;
    rdfs:range :LTCServiceAgreement;
    rdfs:label "delivered by".
:LongTermHealthcareProvider rdfs:subClassOf [
  rdf:type owl:Restriction;
  owl:onProperty :deliveredby;
  owl:qualifiedCardinality "1"^^xsd:nonNegativeInteger;
  owl:onClass :LTCServiceAgreement
].
```

### CODE REVIEW

**Comment by tgoprince** [comment]
There are some issues with the transformation:

- The cardinality constraints of the associations are reversed.

This model:
Schermata 2020-03-01 alle 19 15 15

Is generating this:
Schermata 2020-03-01 alle 19 12 25

While it should be generating this:

Schermata 2020-03-01 alle 19 17 36

- Why are we generating this 99999 cardinality?
Schermata 2020-03-01 alle 19 18 54

If this is because of a cardinality of "*" or "0..*", remember that in those cases, there are no constraints to be generated!

**Comment by tgoprince** [comment]
By the way, I made to small commits to facilitate testing.

**Comment by tgoprince** [comment]
Here are some improvements I believe this feature can benefit from:

* Could you lowerCamelCase names of associations and UpperCamelCase names of classes when generating URIs from them?
* We have to do something about the domain and range of associations and how to transform them. For instance, the object property `gufo:inheresIn` has `gufo:Aspect` as Domain and `gufo:ConcreteIndividual` as Target. However, when you transform the <<characterization>> `reachable by` from the model, we ended up with <<category>> `Agent` in the Domain and <<quality>> `Contact Point` in the range. This creates a logical inconsistency in the generated OWL as, `Agent` is not an `Aspect`. Do you see the problem?
* When an association has no name, we should generate a default name for the association. For instance, the <<componentOf>> between  the classes `Lift` and `Mountain Area`, could become the object property `isComponentOfMountainArea`. In such cases, please also add a `rdfs:comment` informing the name was automatically generated.

**Comment by LucasBassetti** [comment]
@tgoprince I fixed all problems except the domain/range. I will need some help to this one

**Comment by tgoprince** [comment]
Hi @LucasBassetti,

Regarding the URI generation, we still need to make some small adjustments.

We should not use the original OntoUML stereotype names as the prefix, but the gUFO equivalent name.

Consider the generated model below:

Schermata 2020-03-02 alle 10 25 54

The sub properties of categorizes should be:

* `categorizesEvent` instead of `isInstantiationEventType`
* `categorizesLift` instead of `isInstantiationLiftType`
*  `inheresInGeospatialFeature` instead of `isCharacterizationGeospatialFeature`

You already know the mapping from ontouml association stereotypes to gufo object properties, as you use them in the [relation_stereotype_functions.ts](https://github.com/OntoUML/ontouml-js/blob/features/ontouml2gufo-relations/src/libs/ontuml2gufo/relation_stereotype_functions.ts)

File: src/libs/ontuml2gufo/relation_functions.ts [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+    name: classElement.name,
+    uriFormatBy: options.uriFormatBy,
+  });
+
+  const quads = [];
+
+  const lowerboundDomainCardinality = getLowerboundCardinality(cardinality);
+  const upperboundDomainCardinality = getUpperboundCardinality(cardinality);
+  const hasInfiniteCardinality = cardinality.includes('*');
+
```
Comment (tgoprince) [inline src/libs/ontuml2gufo/relation_functions.ts]: You are missing the "1..*" case here.

File: src/libs/ontuml2gufo/relation_functions.ts [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+        writer,
+        relation,
+        isDomain,
+        cardinalityPredicate: 'owl:minQualifiedCardinality',
+        cardinality: lowerboundDomainCardinality,
+        options,
+      }),
+    );
+  }
+  // 2..4, 3..7 ...
```
Comment (tgoprince) [inline src/libs/ontuml2gufo/relation_functions.ts]: This should include the 1...5 case as well.

File: src/libs/ontuml2gufo/relation_functions.ts (hunk 1 above)
Comment (tgoprince) [reply src/libs/ontuml2gufo/relation_functions.ts]: This generates the following snippet:

```
:Marriage rdf:type owl:Class ;
             rdfs:subClassOf gufo:Relator ,
                             [ rdf:type owl:Restriction ;
                               owl:onProperty :mediatesSpouse ;
                               owl:someValuesFrom :Spouse
                             ] ;
```

Comment (tgoprince) [reply src/libs/ontuml2gufo/relation_functions.ts]: Notice that there is no need to use owl:minQualifiedCardinality, as owl:someValuesFrom precisely matches the semantics of 1..*

File: src/libs/ontuml2gufo/relation_stereotype_functions.ts [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+    quad(
+      namedNode(`:${uri}`),
+      namedNode('rdfs:subPropertyOf'),
+      namedNode('gufo:historicallyDependsOn'),
+    ),
+  ];
+}
+
+export function transformInstantiation(
+  relation: IRelation,
```
Comment (tgoprince) [inline src/libs/ontuml2gufo/relation_stereotype_functions.ts]: When transforming this type of relation, you need to **invert** the association ends defined in OntoUML schema.

The reason is that <<instantiation>> has the higher order type as the target and gufo:categorizes has the higher order type as the domain.

File: src/libs/ontuml2gufo/relation_functions.ts (hunk 1 above)
Comment (LucasBassetti) [reply src/libs/ontuml2gufo/relation_functions.ts]: Done

File: src/libs/ontuml2gufo/relation_functions.ts (hunk 2 above)
Comment (LucasBassetti) [reply src/libs/ontuml2gufo/relation_functions.ts]: Done

File: src/libs/ontuml2gufo/relation_stereotype_functions.ts (hunk 3 above)
Comment (LucasBassetti) [reply src/libs/ontuml2gufo/relation_stereotype_functions.ts]: Done

**Pull request merged**